File: hamiguitan_trekscan_plus/lib/models/booking_model.dart
import 'dart:convert';
import 'dart:math';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'member.dart';

class Attachment {
  final String storagePath;
  final String downloadURL;
  final String fileName;
  final String? mimeType;
  final int size;
  final Timestamp uploadedAt;

  Attachment({
    required this.storagePath,
    required this.downloadURL,
    required this.fileName,
    this.mimeType,
    required this.size,
    required this.uploadedAt,
  });

  Map<String, dynamic> toMap() => {
    'storagePath': storagePath,
    'downloadURL': downloadURL,
    'fileName': fileName,
    'mimeType': mimeType,
    'size': size,
    'uploadedAt': uploadedAt,
  };

  factory Attachment.fromMap(Map<String, dynamic> m) => Attachment(
    storagePath: m['storagePath'] as String,
    downloadURL: m['downloadURL'] as String,
    fileName: m['fileName'] as String,
    mimeType: m['mimeType'] as String?,
    size: (m['size'] as num).toInt(),
    uploadedAt: m['uploadedAt'] as Timestamp,
  );
}

class BookingModel {
  final String id; // Now non-nullable - always generated or from Firestore
  final String userId;
  final String affiliation;
  final Timestamp trekDate;
  final String
  trekType; // 'special_trek', 'benchmarking_trek', 'research_trek', 'regular_trek'
  final String hometown; // Primary contact's hometown/city
  final String phoneNumber; // Primary contact's phone number
  final String? notes;
  final String? adminNotes;
  List<Attachment> attachments;
  List<Member> members; // Group members - primary contact is first member
  String status;
  String
  submissionStatus; // 'draft' (saved locally) or 'submitted' (sent to system)
  final Timestamp createdAt;
  Timestamp? updatedAt;

  BookingModel({
    String? id,
    required this.userId,
    required this.affiliation,
    required this.trekDate,
    required this.trekType,
    this.hometown = '',
    this.phoneNumber = '',
    this.notes,
    this.adminNotes,
    this.attachments = const [],
    this.members = const [],
    this.status = 'pending',
    this.submissionStatus = 'draft', // Default to draft when creating
    Timestamp? createdAt,
    this.updatedAt,
  }) : id = id ?? _generateDraftId(),
       createdAt = createdAt ?? Timestamp.now();

  Map<String, dynamic> toMap() {
    final map = <String, dynamic>{
      'id': id, // Always include ID
      'userId': userId,
      'affiliation': affiliation,
      'trekDate': trekDate,
      'trekType': trekType,
      'hometown': hometown,
      'phoneNumber': phoneNumber,
      'notes': notes,
      'adminNotes': adminNotes,
      'attachments': attachments.map((a) => a.toMap()).toList(),
      'members': members.map((m) => m.toMap()).toList(),
      'totalMembers': members.length,
      'status': status,
      'submissionStatus': submissionStatus,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
    return map;
  }

  /// Parse adminNotes from Firestore which might be String, List, or Map
  /// Firebase stores it as JSON string: '[{"id":"...","text":"...","adminName":"..."}]'
  static String? _parseAdminNotes(dynamic adminNotes) {
    if (adminNotes == null) return null;

    // If it's a String, try to parse it as JSON first
    if (adminNotes is String) {
      final trimmed = adminNotes.trim();
      if (trimmed.isEmpty) return null;

      // Check if it's a JSON string (starts with [ or {)
      if (trimmed.startsWith('[') || trimmed.startsWith('{')) {
        try {
          final parsed = jsonDecode(trimmed);
          // Recursively parse the decoded JSON
          return _parseAdminNotes(parsed);
        } catch (e) {
          // If JSON parsing fails, return null to hide malformed data
          return null;
        }
      }
      // Plain text string
      return trimmed;
    }

    // If it's a List (e.g., [{"text": "..."}]), extract first message
    if (adminNotes is List && adminNotes.isNotEmpty) {
      final first = adminNotes.first;
      if (first is Map) {
        // Try common field names for the message (text, message, note, adminNotes)
        final message =
            first['text'] ??
            first['message'] ??
            first['note'] ??
            first['adminNotes'];
        if (message is String && message.trim().isNotEmpty) {
          return message.trim();
        }
      } else if (first is String) {
        return first.trim().isEmpty ? null : first.trim();
      }
    }

    // If it's a Map, extract message field
    if (adminNotes is Map) {
      final message =
          adminNotes['text'] ??
          adminNotes['message'] ??
          adminNotes['note'] ??
          adminNotes['adminNotes'];
      if (message is String && message.trim().isNotEmpty) {
        return message.trim();
      }
    }

    // Fallback: return null to hide unparseable data
    return null;
  }

  /// Generate a unique draft ID for local storage
  /// Format: draft_TIMESTAMP_RANDOM to ensure uniqueness
  static String _generateDraftId() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final random = Random().nextInt(100000);
    return 'draft_${timestamp}_$random';
  }

  factory BookingModel.fromDoc(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return BookingModel(
      id: doc.id,
      userId: data['userId'] as String,
      affiliation: data['affiliation'] as String? ?? '',
      trekDate: data['trekDate'] as Timestamp,
      trekType: data['trekType'] as String? ?? 'regular_trek',
      hometown: data['hometown'] as String? ?? '',
      phoneNumber: data['phoneNumber'] as String? ?? '',
      notes: data['notes'] as String?,
      adminNotes: _parseAdminNotes(data['adminNotes']),
      attachments:
          (data['attachments'] as List<dynamic>?)
              ?.map((e) => Attachment.fromMap(Map<String, dynamic>.from(e)))
              .toList() ??
          [],
      members: const [], // Members loaded separately from subcollection
      status: data['status'] as String? ?? 'pending',
      submissionStatus: data['submissionStatus'] as String? ?? 'submitted',
      createdAt: data['createdAt'] as Timestamp? ?? Timestamp.now(),
      updatedAt: data['updatedAt'] as Timestamp?,
    );
  }

  /// Create BookingModel from JSON data (for local storage)
  factory BookingModel.fromJson(Map<String, dynamic> data) {
    return BookingModel(
      id: data['id'] as String?,
      userId: data['userId'] as String,
      affiliation: data['affiliation'] as String? ?? '',
      trekDate: _parseTimestamp(data['trekDate']),
      trekType: data['trekType'] as String? ?? 'regular_trek',
      hometown: data['hometown'] as String? ?? '',
      phoneNumber: data['phoneNumber'] as String? ?? '',
      notes: data['notes'] as String?,
      adminNotes: _parseAdminNotes(data['adminNotes']),
      attachments:
          (data['attachments'] as List<dynamic>?)
              ?.map((e) => Attachment.fromMap(Map<String, dynamic>.from(e)))
              .toList() ??
          [],
      members:
          (data['members'] as List<dynamic>?)
              ?.map((e) => Member.fromMap(Map<String, dynamic>.from(e)))
              .toList() ??
          [],
      status: data['status'] as String? ?? 'pending',
      submissionStatus: data['submissionStatus'] as String? ?? 'draft',
      createdAt: _parseTimestamp(data['createdAt']),
      updatedAt: data['updatedAt'] != null
          ? _parseTimestamp(data['updatedAt'])
          : null,
    );
  }

  /// Helper to parse Timestamp from various formats
  static Timestamp _parseTimestamp(dynamic value) {
    if (value is Timestamp) {
      return value;
    } else if (value is String) {
      return Timestamp.fromDate(DateTime.parse(value));
    } else if (value is int) {
      return Timestamp.fromDate(DateTime.fromMillisecondsSinceEpoch(value));
    } else if (value is Map<String, dynamic>) {
      // Firestore timestamp format: {_seconds: int, _nanoseconds: int}
      final seconds = value['_seconds'] as int? ?? 0;
      final nanoseconds = value['_nanoseconds'] as int? ?? 0;
      return Timestamp.fromDate(
        DateTime.fromMillisecondsSinceEpoch(
          (seconds * 1000) + (nanoseconds ~/ 1000000),
        ),
      );
    }
    return Timestamp.now();
  }

  /// Get display name for submission status
  String get submissionStatusDisplayName {
    switch (submissionStatus.toLowerCase()) {
      case 'draft':
        return 'Draft (Not Submitted)';
      case 'submitted':
        return 'Submitted';
      default:
        return submissionStatus;
    }
  }

  /// Check if booking is a draft
  bool get isDraft => submissionStatus.toLowerCase() == 'draft';

  /// Create a copy of this booking with modified fields
  BookingModel copyWith({
    String? id,
    String? userId,
    String? affiliation,
    Timestamp? trekDate,
    String? trekType,
    String? hometown,
    String? phoneNumber,
    String? notes,
    String? adminNotes,
    List<Attachment>? attachments,
    List<Member>? members,
    String? status,
    String? submissionStatus,
    Timestamp? createdAt,
    Timestamp? updatedAt,
  }) {
    return BookingModel(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      affiliation: affiliation ?? this.affiliation,
      trekDate: trekDate ?? this.trekDate,
      trekType: trekType ?? this.trekType,
      hometown: hometown ?? this.hometown,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      notes: notes ?? this.notes,
      adminNotes: adminNotes ?? this.adminNotes,
      attachments: attachments ?? this.attachments,
      members: members ?? this.members,
      status: status ?? this.status,
      submissionStatus: submissionStatus ?? this.submissionStatus,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: Admin Dashboard/Mt.-Hamiguitan-TrekScan-Admin-Portal/src/models/BookingModel.js
// Booking Model - represents booking/trek request data
import { Timestamp } from 'firebase/firestore';
import Attachment from './Attachment';

class BookingModel {
  constructor(data = {}) {
    this.id = data.id || null;
    this.userId = data.userId || '';
    this.affiliation = data.affiliation || '';
    this.trekDate = data.trekDate || Timestamp.now();
    this.numberOfPorters = data.numberOfPorters || 0;
    this.trekType = data.trekType || 'general'; // recreational | research
    this.location = data.location || null; // inside_san_isidro | inside_davao_oriental | outside_davao_oriental
    this.phoneNumber = data.phoneNumber || null;
    this.notes = data.notes || null;
    this.adminNotes = data.adminNotes || null;
    this.attachments = data.attachments || [];
    this.status = data.status || 'pending';
    this.createdAt = data.createdAt || Timestamp.now();
    this.updatedAt = data.updatedAt || null;
  }

  // Convert to Firestore-compatible map
  toMap() {
    const map = {
      userId: this.userId,
      affiliation: this.affiliation,
      trekDate: this.trekDate,
      numberOfPorters: this.numberOfPorters,
      trekType: this.trekType,
      status: this.status,
      createdAt: this.createdAt,
      attachments: this.attachments.map(att => 
        att instanceof Attachment ? att.toMap() : Attachment.fromMap(att).toMap()
      )
    };

    if (this.id) {
      map.id = this.id;
    }
    if (this.location !== null && this.location !== undefined) {
      map.location = this.location;
    }
    if (this.phoneNumber) {
      map.phoneNumber = this.phoneNumber;
    }
    if (this.notes) {
      map.notes = this.notes;
    }
    if (this.adminNotes) {
      map.adminNotes = this.adminNotes;
    }
    if (this.updatedAt) {
      map.updatedAt = this.updatedAt;
    }

    return map;
  }

  // Create from Firestore document
  static fromDoc(doc) {
    if (!doc) return null;
    
    const data = doc.data ? doc.data() : doc;
    const docId = doc.id || data.id;

    return new BookingModel({
      id: docId,
      userId: data.userId || '',
      affiliation: data.affiliation || '',
      trekDate: data.trekDate || Timestamp.now(),
      numberOfPorters: typeof data.numberOfPorters === 'number' 
        ? data.numberOfPorters 
        : parseInt(data.numberOfPorters) || 0,
      trekType: data.trekType || 'recreational',
      location: (data.hometown !== null && data.hometown !== undefined) ? data.hometown : 
                ((data.location !== null && data.location !== undefined) ? data.location : null),
      phoneNumber: data.phoneNumber || null,
      notes: data.notes || null,
      adminNotes: data.adminNotes || null,
      attachments: (data.attachments || []).map(att => Attachment.fromMap(att)),
      status: data.status || 'pending',
      createdAt: data.createdAt || Timestamp.now(),
      updatedAt: data.updatedAt || null
    });
  }

  // Create from map (for API responses or manual construction)
  static fromMap(map) {
    if (!map) return null;

    return new BookingModel({
      id: map.id || null,
      userId: map.userId || '',
      affiliation: map.affiliation || '',
      trekDate: map.trekDate || Timestamp.now(),
      numberOfPorters: typeof map.numberOfPorters === 'number'
        ? map.numberOfPorters
        : parseInt(map.numberOfPorters) || 0,
      trekType: map.trekType || 'recreational',
      location: (map.hometown !== null && map.hometown !== undefined) ? map.hometown : 
                ((map.location !== null && map.location !== undefined) ? map.location : null),
      phoneNumber: map.phoneNumber || null,
      notes: map.notes || null,
      adminNotes: map.adminNotes || null,
      attachments: (map.attachments || []).map(att => Attachment.fromMap(att)),
      status: map.status || 'pending',
      createdAt: map.createdAt || Timestamp.now(),
      updatedAt: map.updatedAt || null
    });
  }

  // Convert to JSON (for API responses)
  toJSON() {
    return {
      id: this.id,
      userId: this.userId,
      affiliation: this.affiliation,
      trekDate: this.trekDate?.toDate?.()?.toISOString() || this.trekDate,
      numberOfPorters: this.numberOfPorters,
      trekType: this.trekType,
      location: this.location,
      phoneNumber: this.phoneNumber,
      notes: this.notes,
      adminNotes: this.adminNotes,
      attachments: this.attachments.map(att => 
        att instanceof Attachment ? att.toJSON() : att
      ),
      status: this.status,
      createdAt: this.createdAt?.toDate?.()?.toISOString() || this.createdAt,
      updatedAt: this.updatedAt?.toDate?.()?.toISOString() || this.updatedAt
    };
  }

  // Helper methods
  isPending() {
    return this.status === 'pending';
  }

  isApproved() {
    return this.status === 'approved';
  }

  isRejected() {
    return this.status === 'rejected';
  }

  isCancelled() {
    return this.status === 'cancelled';
  }

  isRecreational() {
    return this.trekType === 'recreational';
  }

  isResearch() {
    return this.trekType === 'research';
  }
}

export default BookingModel;

